java/ql/test/library-tests/dataflow/taint-jackson/Test.java

```diff
@@ -3,50 +3,95 @@
 import java.io.OutputStream;
 import java.io.StringWriter;
 import java.io.Writer;
+import java.util.Iterator;
 
 import com.fasterxml.jackson.core.JsonFactory;
 import com.fasterxml.jackson.core.JsonGenerator;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.ObjectWriter;
+import com.fasterxml.jackson.databind.ObjectReader;
 
 class Test {
+	public static class Potato {
+		private String name;
+
+		private String getName() {
+			return name;
+		}
+	}
+
 	public static String taint() {
 		return "tainted";
 	}
 
+	public static void sink(Object any) {}
+
 	public static void jacksonObjectMapper() throws java.io.FileNotFoundException, java.io.UnsupportedEncodingException {
 		String s = taint();
 		ObjectMapper om = new ObjectMapper();
 		File file = new File("testFile");
 		om.writeValue(file, s);
+		sink(file); //$hasTaintFlow
 		OutputStream out = new FileOutputStream(file);
 		om.writeValue(out, s);
+		sink(file); //$hasTaintFlow
 		Writer writer = new StringWriter();
 		om.writeValue(writer, s);
+		sink(writer); //$hasTaintFlow
 		JsonGenerator generator = new JsonFactory().createGenerator(new StringWriter());
 		om.writeValue(generator, s);
+		sink(generator); //$hasTaintFlow
 		String t = om.writeValueAsString(s);
-		System.out.println(t);
+		sink(t); //$hasTaintFlow
 		byte[] bs = om.writeValueAsBytes(s);
 		String reconstructed = new String(bs, "utf-8");
-		System.out.println(reconstructed);
+		sink(bs); //$hasTaintFlow
+		sink(reconstructed); //$hasTaintFlow
 	}
 
 	public static void jacksonObjectWriter() throws java.io.FileNotFoundException, java.io.UnsupportedEncodingException {
 		String s = taint();
 		ObjectWriter ow = new ObjectWriter();
 		File file = new File("testFile");
 		ow.writeValue(file, s);
+		sink(file); //$hasTaintFlow
 		OutputStream out = new FileOutputStream(file);
 		ow.writeValue(out, s);
+		sink(out); //$hasTaintFlow
 		Writer writer = new StringWriter();
 		ow.writeValue(writer, s);
+		sink(writer); //$hasTaintFlow
 		JsonGenerator generator = new JsonFactory().createGenerator(new StringWriter());
 		ow.writeValue(generator, s);
+		sink(generator); //$hasTaintFlow
 		String t = ow.writeValueAsString(s);
-		System.out.println(t);
+		sink(t); //$hasTaintFlow
 		byte[] bs = ow.writeValueAsBytes(s);
 		String reconstructed = new String(bs, "utf-8");
-		System.out.println(reconstructed);
+		sink(bs); //$hasTaintFlow
+		sink(reconstructed); //$hasTaintFlow
+	}
+
+	public static void jacksonObjectReader() throws java.io.IOException {
+		String s = taint();
+		ObjectMapper om = new ObjectMapper();
+		ObjectReader reader = om.readerFor(Potato.class);
+		sink(reader.readValue(s));  //$hasTaintFlow
+		sink(reader.readValue(s, Potato.class).name);  //$hasTaintFlow
+		sink(reader.readValue(s, Potato.class).getName());  //$hasTaintFlow
```

Comment:
This test fails as does this one:
https://github.com/github/codeql/pull/5814/files#diff-d947c905365a34b846b5ed166177d4087e649ee718455ec77981934a23c62117R94

Any insights into what's going on here would be greatly appreciated.

https://ghsecuritylab.slack.com/archives/CQJU6RN49/p1620056070330400

Reply:
Do you mean it has no result for the `getName()` call?
You are only modeling field access on types which are deserialized using Jackson, but not method accesses. That is probably the reason then.
(Or should the dataflow library be able to detect getter methods on its own?)

Reply:
I was under the impression that data flow would be able to track that the `get` call because it makes a Field Access and that `Field Access` itself is considered an additional taint step.

Reply:
This ended up being resolved by using `DataFlow::getFieldQualifier`.